File: Amazing Bird/Classes/Sprites/ABNormalBird.m
//
//  ABPlayer.m
//  Amazing Bird
//
//  Created by ChristianEnevoldsen on 24/02/14.
//  Copyright (c) 2014 ChristianEnevoldsen. All rights reserved.
//

#import "ABNormalBird.h"
#import "ABPlayer.h"
#import "ABGameScene.h"

#define kAnimIdleKey @"animIdleKey"
#define kAnimMoveKey @"animMoveKey"

@interface ABNormalBird()
@property (nonatomic, weak) ABPlayer *player;
@property (nonatomic) CGFloat velocity;
@end

@implementation ABNormalBird

- (instancetype)initAtPosition:(CGPoint)position withPlayer:(ABPlayer *)player {

    self = [super initWithTexture:[SKTexture textureWithImageNamed:@"bird_anim0"]  position:position];
    if (self) {
		self.player = player;
		self.hitPoints = 1;
		self.maxHitPoints = 3;
		self.movementSpeed = 7;
	}
    return self;
}

- (void)configurePhysicsBody
{
	CGFloat offsetX = self.size.width * self.anchorPoint.x;
	CGFloat offsetY = self.size.height * self.anchorPoint.y;

	CGMutablePathRef path = CGPathCreateMutable();

    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
        CGPathMoveToPoint(path, NULL, 4 - offsetX, 19 - offsetY);
        CGPathAddLineToPoint(path, NULL, 36 - offsetX, 29 - offsetY);
        CGPathAddLineToPoint(path, NULL, 54 - offsetX, 22 - offsetY);
        CGPathAddLineToPoint(path, NULL, 37 - offsetX, 2 - offsetY);
        CGPathAddLineToPoint(path, NULL, 10 - offsetX, 4 - offsetY);
    } else {
        CGPathMoveToPoint(path, NULL, 3 - offsetX, 9 - offsetY);
        CGPathAddLineToPoint(path, NULL, 18 - offsetX, 14 - offsetY);
        CGPathAddLineToPoint(path, NULL, 25 - offsetX, 11 - offsetY);
        CGPathAddLineToPoint(path, NULL, 18 - offsetX, 1 - offsetY);
        CGPathAddLineToPoint(path, NULL, 4 - offsetX, 3 - offsetY);
        CGPathAddLineToPoint(path, NULL, 1 - offsetX, 9 - offsetY);
    }

	CGPathCloseSubpath(path);

	self.physicsBody = [SKPhysicsBody bodyWithPolygonFromPath:path];

    CGPathRelease(path);
	self.physicsBody.dynamic = NO;
    self.physicsBody.mass = 0.01;
    [self reset];

}

- (void)idle {

	[super idle];
	self.status = ABGameCharacterStatusIdle;
	self.physicsBody.dynamic = NO;
	[self animateWithFrames:sharedBirdAnimFrames interval:1.0/32.0 withKey:kAnimIdleKey];
	if ([self actionForKey:@"idleMovementKey"]) {
		return;
	}
	[self runAction:[SKAction repeatActionForever:[SKAction sequence:@[[SKAction moveByX:0 y:-7 duration:1],
																	   [SKAction moveByX:0 y:7 duration:1]]]] withKey:@"idleMovementKey"];
	
}

- (BOOL)paralyzed
{
	 return self.status == ABGameCharacterStatusDead
		 || self.status == ABGameCharacterStatusDying
		 || self.status == ABGameCharacterStatusReviving;
}

- (void)move
{
	if (![self paralyzed]) {
		self.status = ABGameCharacterStatusMoving;
		self.velocity = self.movementSpeed;
		self.physicsBody.dynamic = YES;
		self.physicsBody.velocity = CGVectorMake(0, 0);
        if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
            [self.physicsBody applyImpulse:CGVectorMake(0, 7.0)];
        } else {
            [self.physicsBody applyImpulse:CGVectorMake(0, 3.5)];
        }
        [self animateWithFrames:sharedBirdAnimFrames interval:1.0/45.0 withKey:kAnimMoveKey];
	}
}

- (void)performDeath
{
	self.status = ABGameCharacterStatusDying;
	self.physicsBody.collisionBitMask = 0;
	self.physicsBody.contactTestBitMask = 0;
}

- (void)reset {
	self.physicsBody.categoryBitMask = ABColliderTypeBird;
	self.physicsBody.contactTestBitMask = ABColliderTypeLifeCloud;
	self.physicsBody.collisionBitMask = ABColliderTypeCloud;
	self.physicsBody.allowsRotation = YES;
}

- (void)revive
{
	if ((self.status != ABGameCharacterStatusReviving)) {
		self.status = ABGameCharacterStatusReviving;
		[self runAction:[SKAction moveTo:CGPointMake([self characterScene].size.width/2, self.position.y) duration:0]];
		self.zRotation = 0;
		self.physicsBody.dynamic = NO;
		SKAction *wait = [SKAction waitForDuration:.5];
		SKAction *moveUp = [SKAction moveTo:CGPointMake(self.characterScene.size.width/2,
														self.characterScene.size.height/2) duration:.7];
		SKAction *rotate = [SKAction rotateByAngle:M_PI * 2 duration:.7];
		SKAction *reviveAction = [SKAction sequence:@[wait, [SKAction group:@[moveUp, rotate]]]];
		reviveAction.timingMode = SKActionTimingEaseIn;
		[self runAction:reviveAction completion:^{
			[self idle];
			[self reset];
		}];
	}
}


- (void)update:(NSTimeInterval)delta
{
	if (self.status != ABGameCharacterStatusReviving) {
		if (self.position.y + self.size.height < 0) {
			self.status = ABGameCharacterStatusDead;
		}
	}

}

#pragma mark - Animations and frames

- (void)animateWithFrames:(NSArray *)frames interval:(NSTimeInterval)interval withKey:(NSString *)key
{
	[super animateWithFrames:frames interval:interval withKey:key];
	if (![key isEqualToString:kAnimIdleKey]) {
		//[self removeActionForKey:@"idleMovementKey"];
	}
}

- (void)completedAnimationWithKey:(NSString *)animKey
{
	if ([animKey isEqualToString:kAnimIdleKey]) {
		[self idle];
	}
}

+ (void)loadSharedFrames
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		sharedBirdAnimFrames = [ABGraphicsUtilities loadFramesFromAtlas:@"bird_anim" baseName:@"bird_anim"];
	});
}

static NSArray *sharedBirdAnimFrames = nil;
- (NSArray *)birdAnimFrames {
	return sharedBirdAnimFrames;
}

- (void)addToScene:(ABGameScene *)scene
{
	[scene addNode:self atWorldLayer:ABWorldLayerGameCharacters];
	self.position = CGPointMake(scene.size.width/2, -100);
	[self revive];
}

- (BOOL)applyDamage:(int)damage
{
	BOOL dead = [super applyDamage:damage];
	if (dead) {
		[self performDeath];
	}
	return dead;
}


@end

File: Amazing Bird/Classes/Sprites/ABNormalBird.swift
import SpriteKit

class ABNormalBird: ABGameCharacter {
    private weak var player: ABPlayer?

    private static var sharedFrames: [SKTexture] = []
    private static let isPad = UIDevice.current.userInterfaceIdiom == .pad

    var isParalyzed: Bool { status.isParalyzed }

    // MARK: - Init

    init(position: CGPoint, player: ABPlayer) {
        let texture = SKTextureAtlas(named: "bird_anim").textureNamed("bird_anim0")
        super.init(texture: texture, position: position)
        self.player = player
        hitPoints = 1
        maxHitPoints = 3
        movementSpeed = 7
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) not supported") }

    // MARK: - Physics

    override func configurePhysicsBody() {
        let offsetX = size.width  * anchorPoint.x
        let offsetY = size.height * anchorPoint.y

        let path = CGMutablePath()
        if ABNormalBird.isPad {
            path.move(to:    CGPoint(x:  4 - offsetX, y: 19 - offsetY))
            path.addLine(to: CGPoint(x: 36 - offsetX, y: 29 - offsetY))
            path.addLine(to: CGPoint(x: 54 - offsetX, y: 22 - offsetY))
            path.addLine(to: CGPoint(x: 37 - offsetX, y:  2 - offsetY))
            path.addLine(to: CGPoint(x: 10 - offsetX, y:  4 - offsetY))
        } else {
            path.move(to:    CGPoint(x:  3 - offsetX, y:  9 - offsetY))
            path.addLine(to: CGPoint(x: 18 - offsetX, y: 14 - offsetY))
            path.addLine(to: CGPoint(x: 25 - offsetX, y: 11 - offsetY))
            path.addLine(to: CGPoint(x: 18 - offsetX, y:  1 - offsetY))
            path.addLine(to: CGPoint(x:  4 - offsetX, y:  3 - offsetY))
            path.addLine(to: CGPoint(x:  1 - offsetX, y:  9 - offsetY))
        }
        path.closeSubpath()

        physicsBody = SKPhysicsBody(polygonFrom: path)
        physicsBody?.isDynamic = false
        physicsBody?.mass = 0.01
        reset()
    }

    // MARK: - Actions

    override func idle() {
        super.idle()
        status = .idle
        physicsBody?.isDynamic = false
        animateWithFrames(ABNormalBird.sharedFrames, interval: 1.0 / 32.0, key: "animIdleKey")
        guard action(forKey: "idleMovementKey") == nil else { return }
        let bob = SKAction.repeatForever(SKAction.sequence([
            SKAction.moveBy(x: 0, y: -7, duration: 1),
            SKAction.moveBy(x: 0, y:  7, duration: 1)
        ]))
        run(bob, withKey: "idleMovementKey")
    }

    override func move() {
        guard !isParalyzed else { return }
        status = .moving
        physicsBody?.isDynamic = true
        physicsBody?.velocity = .zero
        let impulse = ABNormalBird.isPad ? CGVector(dx: 0, dy: 7.0) : CGVector(dx: 0, dy: 3.5)
        physicsBody?.applyImpulse(impulse)
        animateWithFrames(ABNormalBird.sharedFrames, interval: 1.0 / 45.0, key: "animMoveKey")
    }

    override func performDeath() {
        status = .dying
        physicsBody?.collisionBitMask  = 0
        physicsBody?.contactTestBitMask = 0
    }

    override func reset() {
        physicsBody?.categoryBitMask    = ABColliderType.bird
        physicsBody?.contactTestBitMask = ABColliderType.cloud
        physicsBody?.collisionBitMask   = ABColliderType.cloud
        physicsBody?.allowsRotation     = true
    }

    override func revive() {
        guard status != .reviving, let scene = characterScene else { return }
        status = .reviving
        run(SKAction.move(to: CGPoint(x: scene.size.width / 2, y: position.y), duration: 0))
        zRotation = 0
        physicsBody?.isDynamic = false

        let center    = CGPoint(x: scene.size.width / 2, y: scene.size.height / 2)
        let moveUp    = SKAction.move(to: center, duration: 0.7)
        moveUp.timingMode = .easeIn
        let rotate    = SKAction.rotate(byAngle: .pi * 2, duration: 0.7)
        let reviveSeq = SKAction.sequence([
            .wait(forDuration: 0.5),
            .group([moveUp, rotate])
        ])
        run(reviveSeq) { [weak self] in
            self?.idle()
            self?.reset()
        }
    }

    override func update(_ delta: TimeInterval) {
        guard status != .reviving else { return }
        if position.y + size.height < 0 {
            status = .dead
        }
    }

    // MARK: - Damage

    @discardableResult
    override func applyDamage(_ damage: Int) -> Bool {
        let dead = super.applyDamage(damage)
        if dead { performDeath() }
        return dead
    }

    // MARK: - Animation callbacks

    override func completedAnimation(key: String) {
        if key == "animIdleKey" { idle() }
    }

    // MARK: - Scene

    override func addToScene(_ scene: ABGameScene, atPosition position: CGPoint) {
        scene.addNode(self, atWorldLayer: .gameCharacters)
        self.position = CGPoint(x: scene.size.width / 2, y: -100)
        revive()
    }

    // MARK: - Shared frames

    override static func loadSharedFrames() {
        guard sharedFrames.isEmpty else { return }
        let atlas = SKTextureAtlas(named: "bird_anim")
        let count = atlas.textureNames.count / 4
        sharedFrames = (0..<count).map { atlas.textureNamed("bird_anim\($0).png") }
    }
}
